plugins/org.eclipse.n4js.jsdoc/src/org/eclipse/n4js/jsdoc/JSDoc2HoverSerializer.java

```diff
@@ -0,0 +1,287 @@
+			}
+			serializer.md2HtmlBuilder.append("</li>");
+		}
+
+		@Override
+		void close(ITagDefinition tagDef, JSDoc2HoverSerializer serializer) {
+			serializer.md2HtmlBuilder.append("</ul></dd>");
+		}
+	};
+
+	static Map<ITagDefinition, TagHandler> LINETAG_HANDLERS;
+
+	static {
+		LINETAG_HANDLERS = new HashMap<>();
+		LINETAG_HANDLERS.put(N4JSDocletParser.TAG_SPEC, TAG_SECTION);
+		LINETAG_HANDLERS.put(N4JSDocletParser.TAG_AUTHOR, TAG_SECTION);
+		LINETAG_HANDLERS.put(N4JSDocletParser.TAG_TODO, TAG_LIST);
+		LINETAG_HANDLERS.put(N4JSDocletParser.TAG_PARAM, TAG_LIST);
+		LINETAG_HANDLERS.put(N4JSDocletParser.TAG_RETURN, TAG_SECTION);
+		LINETAG_HANDLERS.put(N4JSDocletParser.TAG_SPECFROMDESCR, TAG_IGNORE);
+		LINETAG_HANDLERS.put(N4JSDocletParser.TAG_TASK, TAG_IGNORE);
+		LINETAG_HANDLERS.put(N4JSDocletParser.TAG_REQID, TAG_IGNORE);
+		LINETAG_HANDLERS.put(N4JSDocletParser.TAG_API_NOTE, TAG_SECTION);
+		LINETAG_HANDLERS.put(N4JSDocletParser.TAG_API_STATE, TAG_IGNORE);
+		LINETAG_HANDLERS.put(N4JSDocletParser.TAG_TESTEE, TAG_IGNORE);
+		LINETAG_HANDLERS.put(N4JSDocletParser.TAG_TESTEEFROMTYPE, TAG_IGNORE);
+		LINETAG_HANDLERS.put(N4JSDocletParser.TAG_TESTEETYPE, TAG_IGNORE);
+		LINETAG_HANDLERS.put(N4JSDocletParser.TAG_TESTEEMEMBER, TAG_IGNORE);
+	}
+
+	/**
+	 * new line
+	 *
+	 * @author x
+	 */
+	public final static String NL = "\n * ";
+
+	/**
+	 * prefix of new line in multi-line comments (does not apply to first&last markers)
+	 */
+	public final static String NL_PREFIX = " * ";
+
+	MD2HTMLConvertingBuilder md2HtmlBuilder = new MD2HTMLConvertingBuilder();
+
+	/**
+	 * serialize given {@link EObject} as string representing JSDoc
+	 */
+	public static String toJSDocString(EObject obj) {
+		JSDoc2HoverSerializer serializer = new JSDoc2HoverSerializer();
+		if (obj == null) {
+			return "";
+		}
+		serializer.doSwitch(obj);
+		return serializer.md2HtmlBuilder.done();
+	}
+
+	@Override
+	public Boolean caseDoclet(Doclet doclet) {
+		md2HtmlBuilder.resetMarkdownConverter();
+		appendContents(doclet.getContents());
+
+		Map<ITagDefinition, List<LineTag>> groupedLineTags = new HashMap<>();
+		for (LineTag lineTag : doclet.getLineTags()) {
+			ITagDefinition tagDef = lineTag.getTagDefinition();
+			List<LineTag> group = groupedLineTags.get(tagDef);
+			if (group == null) {
+				group = new ArrayList<>(5);
+				groupedLineTags.put(tagDef, group);
+			}
+			group.add(lineTag);
+		}
+		if (!groupedLineTags.isEmpty()) {
+			md2HtmlBuilder.append("\n<dl>");
+			for (ITagDefinition tagDef : N4JSDocletParser.N4JS_LINE_TAGS) {
+				TagHandler tagHandler = LINETAG_HANDLERS.get(tagDef);
+				if (tagHandler == null) {
+					tagHandler = TAG_SECTION; // default tag handler
+				}
+				if (tagHandler != null) {
+					List<LineTag> lineTags = groupedLineTags.get(tagDef);
+					if (lineTags != null && !lineTags.isEmpty()) {
+						tagHandler.open(tagDef, this);
+						boolean first = true;
+						for (LineTag lineTag : lineTags) {
```

Comment:
at this point, the code gets difficult to understand (due to deep control flow structure)